client/lib/wp/sync-handler/index.js

```diff
@@ -0,0 +1,210 @@
+/**
+ * Module dependencies
+ */
+import config from 'config';
+import localforage from 'localforage';
+import Hashes from 'jshashes';
+import { isWhitelisted } from './whitelist-handler';
+
+import debugFactory from 'debug';
+const debug = debugFactory( 'calypso:sync-handler' );
+
+// defaults
+const defaults = config( 'sync-handler-defaults' );
+defaults.driver = localforage[ defaults.driver ];
+
+/**
+ * SyncHandler class
+ */
+export class SyncHandler {
+	/**
+	 * Create a SyncHandler instance
+	 *
+	 * @param {Object} [setup] - localforage sync setup
+	 * @param {Function} handler - wpcom handler
+	 *
+	 * @return {Function} sync-handler wrapper
+	 */
+	constructor( setup, handler ) {
+		if ( typeof setup === 'function' ) {
+			handler = setup;
+			setup = {};
+		}
+
+		this.setup = Object.assign( {}, defaults, setup );
+		this.reqHandler = handler;
+		return this.syncHandlerInstance( handler );
+	}
+
+	syncHandlerInstance( handler ) {
+		return ( params, callback ) => {
+			// detect and no-sync proxy connection request
+			if ( params.metaAPI && params.metaAPI.accessAllUsersBlogs ) {
+				debug( 'skip - non-sync -proxy-handler request detected' );
+				return this.reqHandler( params, callback );
+			}
+
+			const reqParams = Object.assign( {}, params );
+			const { path } = reqParams;
+
+			// whitelist barrier
+			if ( ! isWhitelisted( params ) ) {
+				debug( 'not whitelisted: skip %o request', path );
+				return handler( params, callback );
+			};
+
+			// response sent flag
+			let responseSent = false;
```

Comment:
Instead of creating a closure, if we only want the callback to be called once, we may be able to wrap this like so: `_.once(callback)` This will let us pull these functions out of syncHandlerInstance.